app/scripts/lib/multichain-api/caip-permission-adapter-middleware.test.js

```diff
@@ -0,0 +1,175 @@
+        },
+        unhandled: {
+          methods: ['foobar'],
+          notifications: [],
+        },
+      },
+      isMultichainOrigin: true,
+    },
+  });
+  const getNetworkConfigurationByNetworkClientId = jest
+    .fn()
+    .mockImplementation((networkClientId) => {
+      const chainId =
+        {
+          mainnet: '0x1',
+          goerli: '0x5',
+        }[networkClientId] || '0x999';
+      return {
+        chainId,
+      };
+    });
+  const handler = (request) =>
+    CaipPermissionAdapterMiddleware(request, {}, next, end, {
+      getCaveat,
+      getNetworkConfigurationByNetworkClientId,
+    });
+
+  return {
+    next,
+    end,
+    getCaveat,
+    getNetworkConfigurationByNetworkClientId,
+    handler,
+  };
+};
+
+describe('CaipPermissionAdapterMiddleware', () => {
+  describe('BARAD_DUR feature flag is not set', () => {
+    beforeAll(() => {
+      delete process.env.BARAD_DUR;
+    });
+
+    it('allows the request when BARAD_DUR feature flag is not set', async () => {
+      const { handler, next } = createMockedHandler();
+      await handler(baseRequest);
+      expect(next).toHaveBeenCalled();
+    });
+
+    it('does not read the permission state', async () => {
+      const { handler, getCaveat } = createMockedHandler();
+      await handler(baseRequest);
+      expect(getCaveat).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('BARAD_DUR feature flag is set', () => {
+    beforeAll(() => {
+      process.env.BARAD_DUR = 1;
+    });
+
+    it('gets the authorized scopes from the CAIP-25 endowement permission', async () => {
+      const { handler, getCaveat } = createMockedHandler();
+      await handler(baseRequest);
+      expect(getCaveat).toHaveBeenCalledWith(
+        'http://test.com',
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    });
+
+    it('allows the request when there is no CAIP-25 endowement permission', async () => {
+      const { handler, getCaveat, next } = createMockedHandler();
+      getCaveat.mockReturnValue(null);
+      await handler(baseRequest);
+      expect(next).toHaveBeenCalled();
+    });
+
+    it('allows the request when the CAIP-25 endowement permission was not granted from the multichain flow', async () => {
+      const { handler, getCaveat, next } = createMockedHandler();
+      getCaveat.mockReturnValue({
+        value: {
+          isMultichainOrigin: false,
+        },
+      });
+      await handler(baseRequest);
+      expect(next).toHaveBeenCalled();
+    });
+
+    it('gets the chainId for the request networkClientId', async () => {
+      const { handler, getNetworkConfigurationByNetworkClientId } =
+        createMockedHandler();
+      await handler(baseRequest);
+      expect(getNetworkConfigurationByNetworkClientId).toHaveBeenCalledWith(
+        'mainnet',
+      );
+    });
+
+    it('throws an error if the requested scope method is not authorized in either the current scope or the wallet scope', async () => {
```

Comment:
total nit but these could probably be in a describe block
`when the CAIP-25 endowment permission was granted over the multichain API`

Reply:
and maybe
```suggestion
    it('throws an error if the requested method is not authorized for the scope specified in the request or the wallet scope', async () => {
```

Reply:
both done here https://github.com/MetaMask/metamask-extension/pull/26296/commits/9ad91514473ef2c396016f66dd3b94fcad383936